File: binding/include/libqpluginloader.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qpluginloader.html
pub const qpluginloader = struct {

    /// New constructs a new QPluginLoader object.
    pub fn New() QtC.QPluginLoader {
        return qtc.QPluginLoader_new();
    }


    /// New2 constructs a new QPluginLoader object.
    pub fn New2(fileName: []const u8) QtC.QPluginLoader {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QPluginLoader_new2(fileName_str);
    }


    /// New3 constructs a new QPluginLoader object.
    pub fn New3(parent: ?*anyopaque) QtC.QPluginLoader {
        return qtc.QPluginLoader_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QPluginLoader object.
    pub fn New4(fileName: []const u8, parent: ?*anyopaque) QtC.QPluginLoader {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QPluginLoader_new4(fileName_str, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPluginLoader_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpluginloader.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Instance(self: ?*anyopaque, ) QtC.QObject {
        return qtc.QPluginLoader_Instance(@ptrCast(self));
    }

    pub fn MetaData(self: ?*anyopaque, ) QtC.QJsonObject {
        return qtc.QPluginLoader_MetaData(@ptrCast(self));
    }

    pub fn StaticInstances() []const u8 {
        return qtc.QPluginLoader_StaticInstances();
    }

    pub fn StaticPlugins() []const u8 {
        return qtc.QPluginLoader_StaticPlugins();
    }

    pub fn Load(self: ?*anyopaque, ) bool {
        return qtc.QPluginLoader_Load(@ptrCast(self));
    }

    pub fn Unload(self: ?*anyopaque, ) bool {
        return qtc.QPluginLoader_Unload(@ptrCast(self));
    }

    pub fn IsLoaded(self: ?*anyopaque, ) bool {
        return qtc.QPluginLoader_IsLoaded(@ptrCast(self));
    }

    pub fn SetFileName(self: ?*anyopaque, fileName: []const u8) void {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
qtc.QPluginLoader_SetFileName(@ptrCast(self), fileName_str);
    }

    pub fn FileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPluginLoader_FileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpluginloader.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPluginLoader_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpluginloader.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetLoadHints(self: ?*anyopaque, loadHints: i32) void {
        qtc.QPluginLoader_SetLoadHints(@ptrCast(self), @intCast(loadHints));
    }

    pub fn LoadHints(self: ?*anyopaque, ) i32 {
        return qtc.QPluginLoader_LoadHints(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPluginLoader_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpluginloader.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPluginLoader_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpluginloader.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPluginLoader_Delete(@ptrCast(self));
    }
};

File: binding/src/libqpluginloader.cpp
#include <QJsonObject>
#include <QObject>
#include <QPluginLoader>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qpluginloader.h>
#include "libqpluginloader.h"
#include "libqpluginloader.hxx"

QPluginLoader* QPluginLoader_new() {
	 return new QPluginLoader();
}

QPluginLoader* QPluginLoader_new2(const libqt_string fileName) {
	 return new QPluginLoader(QString::fromUtf8(fileName.data, fileName.len));
}

QPluginLoader* QPluginLoader_new3(QObject* parent) {
	 return new QPluginLoader(parent);
}

QPluginLoader* QPluginLoader_new4(const libqt_string fileName, QObject* parent) {
	 return new QPluginLoader(QString::fromUtf8(fileName.data, fileName.len), parent);
}

libqt_string QPluginLoader_Tr(const char* s) {
	QString _ret = QPluginLoader::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QObject* QPluginLoader_Instance(QPluginLoader* self) {
	return self->instance();
}

QJsonObject* QPluginLoader_MetaData(const QPluginLoader* self) {
	return new QJsonObject(self->metaData());
}

libqt_list QPluginLoader_StaticInstances() {
	auto _ret = QPluginLoader::staticInstances();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		_data[_i] = _ret[_i];
	}
	return _arr;
}

libqt_list QPluginLoader_StaticPlugins() {
	auto _ret = QPluginLoader::staticPlugins();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

bool QPluginLoader_Load(QPluginLoader* self) {
	return self->load();
}

bool QPluginLoader_Unload(QPluginLoader* self) {
	return self->unload();
}

bool QPluginLoader_IsLoaded(const QPluginLoader* self) {
	return self->isLoaded();
}

void QPluginLoader_SetFileName(QPluginLoader* self, const libqt_string fileName) {
	self->setFileName(QString::fromUtf8(fileName.data, fileName.len));
}

libqt_string QPluginLoader_FileName(const QPluginLoader* self) {
	QString _ret = self->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPluginLoader_ErrorString(const QPluginLoader* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPluginLoader_SetLoadHints(QPluginLoader* self, int loadHints) {
	self->setLoadHints(static_cast<QFlags<QLibrary::LoadHint>>(loadHints));
}

int QPluginLoader_LoadHints(const QPluginLoader* self) {
	return self->loadHints();
}

libqt_string QPluginLoader_Tr2(const char* s, const char* c) {
	QString _ret = QPluginLoader::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPluginLoader_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPluginLoader::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPluginLoader_Delete(QPluginLoader* self) {
    delete self;
}

